### ISSUE TITLE
Add hook to add/remove items from a filebrowser's contextMenu

### ISSUE THREAD

**Issue by nross83** [jupyterlab/jupyterlab#3994]
I am building a custom extension and have a filebrowser that I would like to add my own contextMenu items to.  I tried a variety of things and eventually asked on Gitter as I couldn't find a way to hook into filebrowser's contextMenu.  @ian-r-rose suggested I log an issue as it isn't supported currently.

Looking at the code, a contextMenu is built dynamically when the `contextmenu` event is dispatched.  The building of the menu happens [here](https://github.com/jupyterlab/jupyterlab/blob/6fc64f1f624f7c8e07e37b82b8f8c974262610e3/packages/filebrowser-extension/src/index.ts#L430).  Ideally it would be nice if I could get a hook into this logic to add/remove menu items from the default context menu immediately before it is displayed.

I would love to submit a PR for this, but would appreciate some input on where this feature might belong in the filebrowser API. I'm still fairly new to JupyterLab (I love it btw!!! 👏 ❤️). With a little guidance I'd be happy to do the leg work.

### Workaround with some caveats
This is able to be worked around by rendering my own contextMenu from scratch when the `contextmenu` event is dispatched.  The only hitch is that I had to use a `capture` event and `preventDefault` so the default JupyterLab filebrowser contextMenu doesn't also render (showing two menus on top of each other).  The other downside is that I essentially have to duplicate the code I linked above, but it works.

**Comment by ian-r-rose** [comment]
Thanks for opening the issue, @nross83. Some additional context:

Most places in JupyterLab use the application-wide context menu, which registers commands based on CSS selectors. For those cases it is relatively straightforward to extend the menu with new items. 

The filebrowser context menu is an exception to this. It dynamically creates the context menu with each right-click. The reason for this is the `Open With` submenu, which needs additional information about the file type that is clicked than is available from the CSS selectors.

I am not sure what the best way forwards for this is, but it would definitely be nice for extensions to be able to modify the filebrowser context menu with their own file operations. Two possibilities come to mind:

1. Expose some kind of auxiliary command registry on the file browser object so that extensions can add commands there. Something like
```typescript
browser.addContextMenuCommand(commandID)
```
where the browser promises to call the command with the clicked-upon filename in the `args` for the command.

2. Do something clever with HTML data attributes so that the population of the `Open With` submenu can happen in the same way as the rest of the context menus. Then the filebrowser would be able use the same application-wide context menu as everything else, and extensions could use it in the normal way. I don't know how feasible this is at the moment...

Thoughts?

**Comment by rcaunt** [comment]
@ian-r-rose we have a similar requirement in that we'd like to remove a command, or change the behaviour of it behind the scenes.

Is there any current way to remove the default filebrowser and replace it with a custom one?

**Comment by ian-r-rose** [comment]
Yes, it is possible to swap out the default file browser. You would have to copy the `@jupyterlab/filebrowser-extension` package, modify it to your purposes, and include that in the bundle instead of the one we distribute.

**Comment by ian-r-rose** [comment]
You should give your new plugin a different ID and name than the default one, and disable the default one. Something like
```
jupyter labextension disable @jupyterlab/filebrowser-extension
jupyter labextension install your-extension
```
should do the trick. You would need nodejs/npm in the environment. This could probably be done as some extra steps in the build for a docker image.

**Comment by jasongrout** [comment]
> You would need nodejs/npm in the environment.

nodejs < 10, FYI

**Comment by rcaunt** [comment]
@ian-r-rose that's no problem for us thankfully as we've extended and added that.  Typical, I tried `uninstall` but not `disable`, thank you!

**Comment by ian-r-rose** [comment]
This is fixed by #5409. Now the filebrowser uses the application context menu, and new items can be added there in the same way as anywhere else in the application. Thanks @telamonian!
